sbi/utils/pyroutils.py

```diff
@@ -33,3 +40,142 @@ def prior():
             transforms[name] = biject_to(fn.support).inv
 
     return transforms
+
+
+class EstimatorDistribution(pyro.distributions.TorchDistribution):
+    """
+    Base class for a conditioned `sbi` estimator wrapped as a Pyro distribution.
+    """
+
+    def __init__(self, estimator: Any, condition: torch.Tensor):
+        self._estimator = estimator
+        condition_shape, event_shape = self.get_condition_and_event_shapes()
+        self._check_condition_shape(condition)
+        self._condition = condition
+        self._condition_reshaped = condition.reshape(-1, *condition_shape)
+        super().__init__(
+            batch_shape=condition.shape[: -len(condition_shape)],
+            event_shape=event_shape,
+        )
+
+    @abstractmethod
+    def get_condition_and_event_shapes(self) -> Tuple[torch.Size, torch.Size]:
+        pass
+
+    @property
+    def support(self):
+        # TODO: check if we can infer a more specific support from the estimator
+        return constraints.real  # Assume continuous values
+
+    @property
+    def estimator(self):
+        """The conditioned estimator."""
+        return self._estimator
+
+    @property
+    def condition(self):
+        """The conditioning parameter."""
+        return self._condition
+
+    def _check_condition_shape(self, condition: torch.Tensor):
+        """Check that the shape of `condition` is compatible with the estimator."""
+        condition_shape, _ = self.get_condition_and_event_shapes()
+        if len(condition.shape) < len(condition_shape):
+            raise ValueError(
+                "Dimensionality of condition is too small and does not match the "
+                f"expected dimensionality {len(condition_shape)}. "
+                "It should be compatible with condition_shape "
+                f"{condition_shape}."
+            )
+        else:
+            condition_shape = condition.shape[-len(condition_shape) :]
+            if condition_shape != condition_shape:
+                raise ValueError(
+                    f"Condition shape {condition_shape} is not compatible with "
+                    f"estimator condition shape {condition_shape}"
+                )
+
+    def expand(self, batch_shape: torch.Size, _instance=None):
+        """Expand the batch shape of the distribution."""
+        # because batch shape of the distribution is entirely determined by the
+        # condition, we only need to expand the condition
+        condition_shape, _ = self.get_condition_and_event_shapes()
+        condition = self.condition.expand(*batch_shape, *condition_shape)
+        return type(self)(self.estimator, condition)
```

Comment:
why do we need the `type(self)` here?

Reply:
If we call `dist.expand`, we want it to return an object of the same class as `dist`, not an instance of the parent. Calling `type(self)` (or `self.__class__`) ensures this happens (assuming that the parent class and all children share the same constructor pattern. I'm not certain if this is the best way to do this, but it's the only one I know if other than duplicating `expand` for both children.

Reply:
Because `expand` is only 3 LOC, I've copied it to both children and deleted this implementation from the parent.